src/renderers/dom/fiber/ReactDOMFiber.js

```diff
@@ -359,6 +364,65 @@ var DOMRenderer = ReactFiberReconciler({
     parentInstance.removeChild(child);
   },
 
+  canHydrateInstance(
+    instance: Instance | TextInstance,
+    type: string,
+    props: Props,
+  ): boolean {
+    return instance.nodeType === 1 && type === instance.nodeName.toLowerCase();
+  },
+
+  canHydrateTextInstance(instance: Instance | TextInstance): boolean {
+    return instance.nodeType === 3;
+  },
+
+  getNextHydratableSibling(
+    instance: Instance | TextInstance,
+  ): null | Instance | TextInstance {
+    let node = instance.nextSibling;
+    // Skip non-hydratable nodes.
+    /*
+    while (node && node.nodeType !== 1 && node.nodeType !== 3) {
+      node = node.nextSibling;
+    }
+    */
+    return (node: any);
+  },
+
+  getFirstHydratableChild(
+    parentInstance: Container | Instance,
+  ): null | Instance | TextInstance {
+    let next = parentInstance.firstChild;
+    // Skip non-hydratable nodes.
+    /*
+    while (next && next.nodeType !== 1 && next.nodeType !== 3) {
+      next = next.nextSibling;
+    }
+    */
+    return (next: any);
+  },
+
+  hydrateInstance(
+    instance: Instance,
+    type: string,
+    props: Props,
+    rootContainerInstance: Container,
+    internalInstanceHandle: Object,
+  ): void {
+    precacheFiberNode(internalInstanceHandle, instance);
+    // TODO: Possibly defer this until the commit phase where all the events
+    // get attached.
+    updateFiberProps(instance, props);
+    setInitialProperties(instance, type, props, rootContainerInstance);
```

Comment:
That's a non-goal. We might start warning about server-rendering controlled components.

We could potentially always fire a change event after reviving with the new value.

However, maybe it's better to model it as uncontrolled first and then switch to controlled as you're reviving with the knowledge that you'll have to pick up the current value and do something useful with it.

The first solution is certainly more convenient, but is that a good user experience? It seems to me that most use cases for this would suffer in some way unless you designed both experiences.